wolvic/browser/vr/wvr_manager.cc

```diff
@@ -502,12 +502,75 @@ void WvrManager::DrawFrameSubmitNow(device::WebXrFrame* processing_frame) {
   // Renderer is waiting for the previous frame to render, unblock it now.
   submit_client_->OnSubmitFrameRendered();
 
-  if (webxr_.HaveRenderingFrame())
-    webxr_.EndFrameRendering();
+  if (webxr_.HaveRenderingFrame()) {
+    // It's possible, though unlikely, that the previous rendering frame hasn't
+    // finished yet, for example if an unusually slow frame is followed by an
+    // unusually quick one. In that case, wait for that frame to finish
+    // rendering first before proceeding with this one. The state machine
+    // doesn't permit two frames to be in rendering state at once. (Also, adding
+    // even more GPU work in that condition would be counterproductive.)
+    DVLOG(2) << __func__ << ": wait for previous rendering frame to complete";
+    FinishRenderingFrame();
+  }
+
+  CHECK(!webxr_.HaveRenderingFrame());
+  CHECK(webxr_.HaveProcessingFrame());
+  auto* frame = webxr_.GetProcessingFrame();
+  frame->time_copied = base::TimeTicks::Now();
+
+  frame->render_completion_fence = nullptr;
   webxr_.TransitionFrameProcessingToRendering();
 
-  // See if we can animate a new WebXR frame.
-  WebXrTryStartAnimatingFrame();
+  // We finished processing a frame, unblock a potentially waiting next frame.
+  webxr_.TryDeferredProcessing();
+
+  FinishFrame(processing_frame->index);
+
+  if (submit_client_) {
+    // Create a local GpuFence and pass it to the Renderer via IPC.
+    std::unique_ptr<gl::GLFence> gl_fence = gl::GLFence::CreateForGpuFence();
+    std::unique_ptr<gfx::GpuFence> gpu_fence2 = gl_fence->GetGpuFence();
+    submit_client_->OnSubmitFrameGpuFence(
+        gpu_fence2->GetGpuFenceHandle().Clone());
+  }
+
+  if (pending_getframedata_) {
+    std::move(pending_getframedata_).Run();
+  }
+}
+
+void WvrManager::FinishRenderingFrame() {
+  DCHECK(webxr_.HaveRenderingFrame());
+  device::WebXrFrame* frame = webxr_.GetRenderingFrame();
+  DVLOG(3) << __func__ << ": frame=" << frame->index;
+
+  if (!frame->render_completion_fence) {
+    frame->render_completion_fence = gl::GLFence::CreateForGpuFence();
+  }
+
+  // ClearRenderingFrame
+
+  // Ensure that we're totally finished with the rendering frame, then collect
+  // stats and move the frame out of the rendering path.
+  DVLOG(2) << __func__ << ": client wait start";
+  frame->render_completion_fence->ClientWait();
+  DVLOG(2) << __func__ << ": client wait done";
+
+  webxr_.EndFrameRendering(frame);
+}
+
+void WvrManager::FinishFrame(int16_t frame_index) {
+  DVLOG(2) << __func__;
+
+  // TODO(tiago): Swap Buffers?
```

Comment:
My guess is that it should be done in Wolvic side as it is processing the scenegraph? Where are we getting the buffer to render to? How we cycle them?